main/src/ca/uwaterloo/flix/language/phase/unification/CofiniteIntSet.scala

```diff
@@ -0,0 +1,137 @@
+  def isUniverse: Boolean = this match {
+    case Compl(s) if s.isEmpty => true
+    case Set(_) => false
+    case Compl(_) => false
+  }
+
+  /** Returns `true` if `this` is [[CofiniteIntSet.empty]]. */
+  def isEmpty: Boolean = this match {
+    case Set(s) if s.isEmpty => true
+    case Set(_) => false
+    case Compl(_) => false
+  }
+
+}
+
+object CofiniteIntSet {
+
+  /** Represents a finite set of integers. */
+  private case class Set(s: SortedSet[Int]) extends CofiniteIntSet
+
+  /** Represents a co-finite set of integers. */
+  private case class Compl(s: SortedSet[Int]) extends CofiniteIntSet
+
+  /** The empty set. */
+  val empty: CofiniteIntSet = Set(SortedSet.empty)
+
+  /** The universe set. */
+  val universe: CofiniteIntSet = Compl(SortedSet.empty)
+
+  /** Returns the wrapped set of `s`. */
+  def mkSet(s: SortedSet[Int]): CofiniteIntSet = Set(s)
+
+  /** Returns the singleton set of `i`. */
+  def mkSet(i: Int): CofiniteIntSet = Set(SortedSet(i))
+
+  /** Returns the union of `s1` and `s2` (`s1 ∪ s2`). */
+  def union(s1: CofiniteIntSet, s2: CofiniteIntSet): CofiniteIntSet = (s1, s2) match {
+    case (Set(x), Set(y)) =>
+      // x ∪ y
+      Set(x.union(y))
+    case (Set(x), Compl(y)) =>
+      // x ∪ !y
+      // = !!(x ∪ !y)    (double complement)
+      // = !(!x ∩ y)     (complement distribution)
+      // = !(y ∩ !x)     (intersection symmetry)
+      // = !(y - x)      (difference definition)
+      Compl(y.diff(x))
+    case (Compl(x), Set(y)) =>
+      // !x ∪ y
+      // = !!(!x ∪ y)    (double complement)
+      // = !(x ∩ !y)     (complement distribution)
+      // = !(x - y)      (difference definition)
+      Compl(x.diff(y))
+    case (Compl(x), Compl(y)) =>
+      // !x ∪ !y
+      // = !!(!x ∪ !y)   (double complement)
+      // = !(!!x ∩ !!y)  (complement distribution)
+      // = !(x ∩ y)      (double complement)
+      Compl(x.intersect(y))
+  }
+
+  /** Returns the intersection of `s1` and `s2` (`s1 ∩ s2`). */
+  def intersection(s1: CofiniteIntSet, s2: CofiniteIntSet): CofiniteIntSet = (s1, s2) match {
+    case (Set(x), Set(y)) =>
+      // x ∩ y
+      Set(x.intersect(y))
+    case (Set(x), Compl(y)) =>
+      // x ∩ !y
+      // = x - y         (difference definition)
+      Set(x.diff(y))
+    case (Compl(x), Set(y)) =>
+      // !x ∩ y
+      // = y ∩ !x        (intersection symmetry)
+      // = y - x         (difference definition)
+      Set(y.diff(x))
+    case (Compl(x), Compl(y)) =>
+      // !x ∩ !y
+      // = !!(!x ∩ !y)   (double complement)
+      // = !(x ∪ y)      (complement distribution)
+      Compl(x.union(y))
+  }
+
+  /** Returns the complement of `s` (`!s`). */
+  def complement(s: CofiniteIntSet): CofiniteIntSet = s match {
+    case Set(s) =>
+      // !s
+      Compl(s)
+    case Compl(s) =>
+      // !!s
+      // = s             (double negation)
+      Set(s)
+  }
+
+  /**
+    * Instantiates the elements of `s` according to `m`.
```

Comment:
I don't fully understand what this does.

First I thought it was just map, but its not.

I think an example in the comment would be nice.

Reply:
I will just not include it in this PR, it doesn't make sense without the external context